File: supabase/functions/_shared/embeddings.ts
// Embedding generation using Gemini text-embedding-004

export interface EmbeddingResult {
  embedding: number[];
}

// Generate embedding for a single text (document content)
export async function generateEmbedding(
  text: string,
  apiKey: string
): Promise<EmbeddingResult> {
  const response = await fetch(
    `https://generativelanguage.googleapis.com/v1beta/models/text-embedding-004:embedContent?key=${apiKey}`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        model: 'models/text-embedding-004',
        content: {
          parts: [{ text }],
        },
        taskType: 'RETRIEVAL_DOCUMENT',
      }),
    }
  );

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Embedding API error: ${error}`);
  }

  const data = await response.json();
  return {
    embedding: data.embedding.values,
  };
}

// Generate embedding optimized for queries
export async function generateQueryEmbedding(
  query: string,
  apiKey: string
): Promise<EmbeddingResult> {
  const response = await fetch(
    `https://generativelanguage.googleapis.com/v1beta/models/text-embedding-004:embedContent?key=${apiKey}`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        model: 'models/text-embedding-004',
        content: {
          parts: [{ text: query }],
        },
        taskType: 'RETRIEVAL_QUERY',
      }),
    }
  );

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Query embedding API error: ${error}`);
  }

  const data = await response.json();
  return {
    embedding: data.embedding.values,
  };
}

// Batch embedding for efficiency (up to 100 texts per call)
export async function generateEmbeddingsBatch(
  texts: string[],
  apiKey: string
): Promise<EmbeddingResult[]> {
  if (texts.length === 0) {
    return [];
  }

  // Gemini batch limit is 100
  if (texts.length > 100) {
    // Process in batches of 100
    const results: EmbeddingResult[] = [];
    for (let i = 0; i < texts.length; i += 100) {
      const batch = texts.slice(i, i + 100);
      const batchResults = await generateEmbeddingsBatch(batch, apiKey);
      results.push(...batchResults);
    }
    return results;
  }

  const response = await fetch(
    `https://generativelanguage.googleapis.com/v1beta/models/text-embedding-004:batchEmbedContents?key=${apiKey}`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        requests: texts.map((text) => ({
          model: 'models/text-embedding-004',
          content: {
            parts: [{ text }],
          },
          taskType: 'RETRIEVAL_DOCUMENT',
        })),
      }),
    }
  );

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Batch embedding API error: ${error}`);
  }

  const data = await response.json();
  return data.embeddings.map((e: { values: number[] }) => ({
    embedding: e.values,
  }));
}

File: backend-py/services/embeddings.py
"""Embedding generation using Gemini gemini-embedding-001 (768 dimensions via outputDimensionality)."""

import os
import httpx

EMBEDDING_MODEL = "gemini-embedding-001"
EMBEDDING_DIMS = 768  # Match existing DB vector(768) column


def _api_key() -> str:
    key = os.getenv("GOOGLE_API_KEY")
    if not key:
        raise RuntimeError("GOOGLE_API_KEY is not configured")
    return key


async def generate_embedding(text: str, api_key: str | None = None) -> list[float]:
    """Generate embedding for a single document text (RETRIEVAL_DOCUMENT task)."""
    key = api_key or _api_key()
    url = f"https://generativelanguage.googleapis.com/v1beta/models/{EMBEDDING_MODEL}:embedContent?key={key}"

    payload = {
        "model": f"models/{EMBEDDING_MODEL}",
        "content": {"parts": [{"text": text}]},
        "taskType": "RETRIEVAL_DOCUMENT",
        "outputDimensionality": EMBEDDING_DIMS,
    }

    async with httpx.AsyncClient(timeout=60.0) as client:
        resp = await client.post(url, json=payload)
        if resp.status_code != 200:
            raise RuntimeError(f"Embedding API error: {resp.text}")
        data = resp.json()
        return data["embedding"]["values"]


async def generate_query_embedding(query: str, api_key: str | None = None) -> list[float]:
    """Generate embedding optimized for queries (RETRIEVAL_QUERY task)."""
    key = api_key or _api_key()
    url = f"https://generativelanguage.googleapis.com/v1beta/models/{EMBEDDING_MODEL}:embedContent?key={key}"

    payload = {
        "model": f"models/{EMBEDDING_MODEL}",
        "content": {"parts": [{"text": query}]},
        "taskType": "RETRIEVAL_QUERY",
        "outputDimensionality": EMBEDDING_DIMS,
    }

    async with httpx.AsyncClient(timeout=60.0) as client:
        resp = await client.post(url, json=payload)
        if resp.status_code != 200:
            raise RuntimeError(f"Query embedding API error: {resp.text}")
        data = resp.json()
        return data["embedding"]["values"]


async def generate_embeddings_batch(texts: list[str], api_key: str | None = None) -> list[list[float]]:
    """Batch embedding for efficiency (up to 100 texts per call, recursive for larger)."""
    if not texts:
        return []

    key = api_key or _api_key()

    # Gemini batch limit is 100
    if len(texts) > 100:
        results: list[list[float]] = []
        for i in range(0, len(texts), 100):
            batch = texts[i : i + 100]
            batch_results = await generate_embeddings_batch(batch, key)
            results.extend(batch_results)
        return results

    url = f"https://generativelanguage.googleapis.com/v1beta/models/{EMBEDDING_MODEL}:batchEmbedContents?key={key}"

    payload = {
        "requests": [
            {
                "model": f"models/{EMBEDDING_MODEL}",
                "content": {"parts": [{"text": t}]},
                "taskType": "RETRIEVAL_DOCUMENT",
                "outputDimensionality": EMBEDDING_DIMS,
            }
            for t in texts
        ]
    }

    async with httpx.AsyncClient(timeout=120.0) as client:
        resp = await client.post(url, json=payload)
        if resp.status_code != 200:
            raise RuntimeError(f"Batch embedding API error: {resp.text}")
        data = resp.json()
        return [e["values"] for e in data["embeddings"]]
